google-cloud-storage/src/main/java/com/google/cloud/storage/GrpcStorageImpl.java

```diff
@@ -987,43 +1007,84 @@ public String getNextPageToken() {
 
     @Override
     public Page<ModelT> getNextPage() {
-      return new TransformingPageDecorator<>(page.getNextPage(), translator);
+      return new TransformingPageDecorator<>(
+          page.getNextPage(), translator, deps, resultRetryAlgorithm);
     }
 
+    @SuppressWarnings({"Convert2MethodRef"})
     @Override
     public Iterable<ModelT> iterateAll() {
-      return () -> {
-        final Iterator<ResourceT> iter = page.iterateAll().iterator();
-        return new TransformingIterator(iter);
-      };
+      // iterateAll on AbstractPage isn't very friendly to decoration, as getNextPage isn't actually
+      // ever called. This means we aren't able to apply our retry wrapping there.
+      // Instead, what we do is create a stream which will attempt to call getNextPage repeatedly
+      // until we meet some condition of exhaustion. At that point we can apply our retry logic.
+      return () ->
+          streamIterate(
+                  page,
+                  p -> p != null && p.hasNextPage(),
+                  prev -> {
+                    // explicitly define this callable rather than using the method reference to
+                    // prevent a javac 1.8 exception
+                    // https://bugs.java.com/bugdatabase/view_bug.do?bug_id=8056984
+                    Callable<PageT> c = () -> prev.getNextPage();
+                    return Retrying.run(deps, resultRetryAlgorithm, c, Decoder.identity());
```

Comment:
only question i have is; for exceptions thrown from `Retrying.run`, what happens to them? I ask because I don't see `throws Exception` in the method signatures.

Reply:
Exceptions which aren't retryable and/or exhaust the `Retrying.run` will be caught and coalesced to a StorageException.

https://github.com/googleapis/java-storage/pull/1571/files#diff-16ae9534e54393cc0a423731519677b6d8d1dae39e70afe603579abff529860cR85-R96

Reply:
My question is more about checked exception vs. non-checked exception. In this case it would be unchecked exception but not sure why we wouldn't.

Reply:
The retrying code under the hood will catch any checked exception and wrap it.